src/vivarium/framework/engine.py

```diff
@@ -35,107 +33,139 @@
+        self._lifecycle.add_phase('setup', ['setup', 'post_setup', 'population_creation'])
+        self._lifecycle.add_phase(
+            'main_loop', ['time_step__prepare', 'time_step', 'time_step__cleanup', 'collect_metrics'], loop=True
+        )
+        self._lifecycle.add_phase('simulation_end', ['simulation_end', 'report'])
 
-        self.builder = Builder(configuration, plugin_manager)
+        self._component_manager = self._plugin_manager.get_plugin('component_manager')
+        self._component_manager.setup(self.configuration, self._lifecycle)
 
-        self.component_manager = plugin_manager.get_plugin('component_manager')
-        self.clock = plugin_manager.get_plugin('clock')
-        self.values = plugin_manager.get_plugin('value')
-        self.events = plugin_manager.get_plugin('event')
-        self.population = plugin_manager.get_plugin('population')
-        self.tables = plugin_manager.get_plugin('lookup')
-        self.randomness = plugin_manager.get_plugin('randomness')
-        self.data = plugin_manager.get_plugin('data')
+        self._clock = self._plugin_manager.get_plugin('clock')
+        self._values = self._plugin_manager.get_plugin('value')
+        self._events = self._plugin_manager.get_plugin('event')
+        self._population = self._plugin_manager.get_plugin('population')
+        self._tables = self._plugin_manager.get_plugin('lookup')
+        self._randomness = self._plugin_manager.get_plugin('randomness')
+        self._data = self._plugin_manager.get_plugin('data')
 
-        for name, controller in plugin_manager.get_optional_controllers().items():
-            setattr(self, name, controller)
+        for name, controller in self._plugin_manager.get_optional_controllers().items():
+            setattr(self, f'_name', controller)
 
         # The order the managers are added is important.  It represents the
         # order in which they will be set up.  The clock is required by
         # several of the other managers, including the lifecycle manager.  The
         # lifecycle manager is also required by most managers. The randomness
         # manager requires the population manager.  The remaining managers need
         # no ordering.
-        self.component_manager.add_managers(
-            [self.clock, self.lifecycle, self.population, self.randomness,
-             self.values, self.events, self.tables, self.data]
-        )
-        self.component_manager.add_managers(list(plugin_manager.get_optional_controllers().values()))
-        self.component_manager.add_components(components + [Metrics()])
+        managers = [self._clock, self._lifecycle, self._population, self._randomness, self._values, self._events,
+                    self._tables, self._data] + list(self._plugin_manager.get_optional_controllers().values())
+        self._component_manager.add_managers(managers)
+
+        component_config_parser = self._plugin_manager.get_plugin('component_configuration_parser')
```

Comment:
Pushed component parsing way back into the process.  The above setup is all super snappy, so we'll still fail fast *enough*.